src/qengine/arithmetic.cpp

```diff
@@ -464,32 +462,50 @@ void QEngineCPU::ModNOut(const MFn& kernelFn, const bitCapInt& modN, const bitLe
         bitCapInt otherRes = lcv & otherMask;
         bitCapInt inRes = lcv & inMask;
         bitCapInt outRes = (kernelFn(inRes >> inStart) % modN) << outStart;
-        nStateVec->write(inRes | outRes | otherRes, stateVec->read(lcv));
+        if (inverse) {
+            nStateVec->write(lcv, stateVec->read(inRes | outRes | otherRes));
+        } else {
+            nStateVec->write(inRes | outRes | otherRes, stateVec->read(lcv));
+        }
     });
 
     ResetStateVec(nStateVec);
 }
 
 void QEngineCPU::MULModNOut(bitCapInt toMod, bitCapInt modN, bitLenInt inStart, bitLenInt outStart, bitLenInt length)
 {
+    SetReg(outStart, length, 0);
+
     if (toMod == 0) {
-        SetReg(outStart, length, 0);
         return;
     }
 
     ModNOut([&toMod](const bitCapInt& inInt) { return inInt * toMod; }, modN, inStart, outStart, length);
 }
 
+void QEngineCPU::IMULModNOut(bitCapInt toMod, bitCapInt modN, bitLenInt inStart, bitLenInt outStart, bitLenInt length)
+{
+    if (toMod == 0) {
+        return;
+    }
+
+    ModNOut([&toMod](const bitCapInt& inInt) { return inInt * toMod; }, modN, inStart, outStart, length, true);
+}
+
 void QEngineCPU::POWModNOut(bitCapInt toMod, bitCapInt modN, bitLenInt inStart, bitLenInt outStart, bitLenInt length)
 {
+    if (toMod == ONE_BCI) {
+        SetReg(outStart, length, ONE_BCI);
```

Comment:
In the MULModNOut you pulled this line out of the if statement.  Is that appropriate here, as well?

Reply:
No, it is not. In MULModNOut, the SetReg() needs to happen regardless. Here, we set it to 1 only in this case, otherwise we init it to 0 and proceed from there.